Re-added executable

sblackler opened the pull request:
Here lies one of the last remaining bits for #104 

As mentioned on twitter. There are couple of failing tests that we can discuss below before merging this back in. Details to follow

smudge202 commented:
The tests names are no good. The last thing we need are tests in the state they were before with inconsistencies everywhere.

I don't think the format provided here makes sense/improve over the existing format. The blocks are Given/When/Then and at first glance the tests almost format that. However, in the case of 
`GivenAnExecutableOfTContextTResult`
`WhenExecuteIsInvoked`
`WithNoServicesRegisteredThenThrowsException`
that final method doesn't follow the convention. The `WIthNoServicesRegistered` is a given.

Of course, the existing format would also suffer this, mixing the Given into the service name. As you've probably spotted though, (despite having at one stage provided an `xunit.json`) test explorer displays names of tests in classes like this as `GivenAnExecutableOfTContextTResult...` which provides no information. When we get method naming working again in xunit, it will instead display `WithNoServicesRegisteredThenThrowsException`. Again, there's not enough context there to know what the test is.

Had it been the existing format, the test name in method display mode would be
`GivenNoServicesRegisteredWhenExecuteIsInvokedThenThrowsException`. This provides much better context.

Point being, we can't have inconsistent naming schemes again (which holds true to the generics; everywhere else there's no `TFoo`, just `Foo`).

sblackler commented:
Changed to be all one liners and removed T from the executable class. Think I got the right name on the second class. If you're happy with the naming now, i will adjust my other working branch.

smudge202 commented on src/Compose/Executable.cs:
```diff
@@ -0,0 +1,179 @@
+                ExecutionAsync = asyncInvoke;
+            }
+
+            public void OnExecute<TService>(Action<TService, Result> invoke)
```
I appreciate this is old code but stil have some `T`'s here.

sblackler commented on src/Compose/Executable.cs:
```diff
@@ -0,0 +1,179 @@
+                ExecutionAsync = asyncInvoke;
+            }
+
+            public void OnExecute<Service>(Action<Service, Result> invoke)
```
@smudge202 Shouldn't this have `where Service : class` as per the other implementations?

smudge202 replied:
Not necessary. Everywhere else this is a forced constraint because further down in the stack it wouldn't compile otherwise. I don't know if this is true for executable... perhaps it is?

sblackler replied:
I'm thinking for consistency we add it in?

smudge202 replied:
Personally, I can't think of any example whereby it would be logical to have a value type be DI injected. For such cases, I'd suggest using OptionsModel as I can only assume it be part of the application configuration, not DI.

So yeh, sure, can be added in.

sblackler replied:
Added

sblackler commented:
Updated.

sblackler commented:
So the only test(s) that are failing are these:

```c#
        [Unit]
        public void GivenAnExecutableOfResultWhenOnExecuteIsInvokedAndNoServicesRegisteredThenThrowsException()
        {
            var app = new Fake.FakeExecutable();

            Action act = () => app.OnExecute<string>((service) => true);

            act.ShouldThrow<InvalidOperationException>();
        }
```

By the looks of things the functionality has changed which now causes this. Do we still want this behavior or not? If so, I will modify and make the tests pass

smudge202 commented:
Given the title of that test, and the state of the repo, I would still expect the test to be viable, but the implementation of said test will need changing to be compilable.

```c#
        [Unit]
        public void GivenAnExecutableOfResultWhenOnExecuteIsInvokedAndNoServicesRegisteredThenThrowsException()
        {
            var app = new Fake.FakeExecutable();

            Action act = () => app.OnExecute(() => true);

            act.ShouldThrow<InvalidOperationException>();
        }
```

The test in it's current form would probably throw an exception (not sure if invalid operation) because it's trying to resolve a string from an empty service collection; not for the reason specified in the test title.

sblackler commented:
Changing the test to make it pass will force the order to be Add Services -> Set OnExecute -> Execute. In my mind, so long as the first two are done before execute it shouldn't matter?  Basing this on the deferred execution stuff you mentioned the other day. Therefore, I'm thinking that these tests are actually mute and can be removed (or changed so that the method under test is execute not OnExecute)

smudge202 commented:
Hmmm. I would prefer that OnExecute/UseServices be in any order come to think of it. And you should be able to Execute without UseServices but OnExecute should throw exception?

sblackler commented:
The exception should come from Execute not OnExecute as that's the point where we say we've done everything and we are okay to run. So not having services there makes no sense.

sblackler commented:
For completeness, on stream we discussed the fact these tests were probably failing incorrectly anyway, so we have decided to remove them.
